contracts/Vault.vy

```diff
@@ -1582,32 +1584,35 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
         / SECS_PER_YEAR
     )
 
-    # Only applies in certain conditions
-    strategist_fee: uint256 = 0
-    performance_fee: uint256 = 0
+    self.managementDebt +=  management_fee
+
+    if gain == 0:
+        return 0
+
+    # if total performance fee is greater than 100% then this will cause an underflow
+    assert self.strategies[strategy].performanceFee + self.performanceFee <= MAX_BPS
 
     # NOTE: Applies if Strategy is not shutting down, or it is but all debt paid off
     # NOTE: No fee is taken when a Strategy is unwinding it's position, until all debt is paid
-    if gain > 0:
-        # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
-        strategist_fee = (
-            gain
-            * self.strategies[strategy].performanceFee
-            / MAX_BPS
-        )
-        # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
-        performance_fee = gain * self.performanceFee / MAX_BPS
+    # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
+    strategist_fee: uint256 = (
+        gain
+        * self.strategies[strategy].performanceFee
+        / MAX_BPS
+    )
+    # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
+    performance_fee: uint256 = gain * self.performanceFee / MAX_BPS
 
     # NOTE: This must be called prior to taking new collateral,
     #       or the calculation will be wrong!
     # NOTE: This must be done at the same time, to ensure the relative
     #       ratio of governance_fee : strategist_fee is kept intact
-    total_fee: uint256 = performance_fee + strategist_fee + management_fee
+    total_fee: uint256 = performance_fee + strategist_fee + self.managementDebt
+    self.managementDebt = 0
     # ensure total_fee is not more than gain
     if total_fee > gain:
+        self.managementDebt = total_fee - gain
         total_fee = gain
-        # if total performance fee is greater than 100% then this will cause an underflow
-        management_fee = gain - performance_fee - strategist_fee
```

Comment:
management_fee is indeed not used after being updated @fubuloubu @Grandthrax 

https://github.com/yearn/yearn-vaults/pull/263/files#diff-b583ba5c4d7b67e56b9084eff9fe883e12a2cdd94a2ea25029322c6f3568d807R1562

Reply:
This is because the total balance is sent to the rewards for the management fee. So indeed, this invariant is kept but it is unused because the resulting value sent to rewards is:
```
_issueSharesForAmount( management_fee + performance_fee )
= _issueSharesForAmount( (gain - performance_fee - strategist_fee) + performance_fee )
= _issueSharesForAmount( gain - strategist_fee )
= token.balanceOf(self)
```

Reply:
https://github.com/yearn/yearn-vaults/blob/cc691c4d725b593b5a9ee406fd6dbe9509435aa8/contracts/Vault.vy#L1605-L1609
when total_fee < gain
exec management_fee may < management_fee
should be add var storage management_fee - (exec management_fee)
https://github.com/yearn/yearn-vaults/blob/cc691c4d725b593b5a9ee406fd6dbe9509435aa8/contracts/Vault.vy#L1575-L1583

Reply:
Split this into two PRs:
1. The above 2 lines removed, with no other changes (with the documented reasoning that there is no need for this variable do to the implicit invariant that all leftover minted shares are sent to `rewards`)
2. This new "fix" (which we are having trouble parsing) using `managementDebt` and the associated updates to Vault and test cases